rfcs/000-remove-double-relationships.md

```diff
@@ -0,0 +1,103 @@
+- Feature Name: Remove double relationships between Neo4j nodes
+- Start Date: 2021-05-25
+- RFC PR: [amundsen-io/rfcs#0000](https://github.com/amundsen-io/rfcs/pull/0000)
+- Amundsen Issue: 
+
+# Remove double relationships between Neo4j nodes
+
+## Summary
+
+Currently all (most?) relationships between Neo4j nodes are modeled with
+two directional edges. This RFC proposes we only use one directional edge
+between the nodes.
+
+## Motivation
+
+The double relationship graph model makes Neo4j queries more complicated
+than they need to be, with no performance or functional benefit. Additionally
+due to conflicting changes in the past ([PR 70](https://github.com/amundsen-io/amundsenmetadatalibrary/pull/70)
+and [PR 206](https://github.com/amundsen-io/amundsenmetadatalibrary/pull/206)),
+Amundsen instances that have been running and ingesting data since prior to
+PR 70 or PR 206 have inconsistent Neo4j database state. Some nodes have a
+single relationship between then, some have two.
+
+There are two relevant Slack discussions regarding this change:
+
+https://app.slack.com/client/TGFR0CZM3/CGFBVT23V/thread/CGFBVT23V-1615391140.022900
+- a thread where I asked the question on why the double relationships exist
+https://app.slack.com/client/TGFR0CZM3/CGFBVT23V/thread/CGFBVT23V-1563487877.254000
+- a thread with Joshua Haskins' benchmarking results for modeling the relationships
+  with one or two edges
+
+## Transition Path
+
+The implementation work involved:
+
+For each double relationship pair, we need to decide which relationship type
+to keep and which one to remove. I suggest we do this by keeping the directionality
+and tag name consistent (e.g. remove all relationships with _BY suffix). There
+might be instances where we need to change the direction of the relationship,
+because FOO_BY and BAR_BY relationships might've been defined using the opposite
+directionality.
+
+Delete the insertion of the relationship type we've decided to remove from the
+code, and, if needed, change the direction of the relationships we've decided to
+keep.
+
+Create a database migration script that deletes the relationship types we've
+decided to remove from the Neo4j database, and change the direction of any
+relationship we've decided to keep, if needed.
+
+There is no transition path for new installations. They will simply start using
+the single relationships graph model from the get go.
+
+For existing Amundsen instances we'll need to make sure the database migration
+script is run after upgrading Amundsen to a version where the double relationships
+have been removed.
+
+## How We Communicate This
+
+TBD
```

Comment:
While communication to the community should help to mitigate any issues, it seems like one of the largest potential risks (IMHO) could be having a version mismatch b/w the data (largely impacted by data builder) and the metadata service. In addition, there isn't a hard requirement that links databuilder and metadata today that we could rely on to enforce version compliance.

Do you have thoughts on the feasibility on making the metadata layer backwards compatible and how the rollout of this could be "accident proof" across services to de-risk any implementation. ?

Reply:
I think if we freeze the metadata layer as it is now, and factor out the model classes that are defined separately in databuilder and metada/search into a common library, we can rely on that ancestor as the source of truth, and use that to keep things in sync. A good place to start is amundsenrdns IMHO

Reply:
I'm largely aligned here and, should this approach be taken in the RFC, would like to see the next level of detail on the how to do this in the RFC.

Reply:
I'm submitting some changes to the RFC in a minute to discuss how this already happened in PR 70 referred in the RFC.

Personally I'm of the opinion the problem you're asking about should be addressed separately. It might be beneficial to address it before working on implementing this RFC, though.